File: src/main/java/com/oycm/algorithm/h/basic/Solution_17.java
package com.oycm.algorithm.h.basic;

public class Solution_17 {

    /**
     * 338. <a href="https://leetcode.cn/problems/counting-bits/description/">比特位计数</a>
     *
     * todo 后续尝试动态规划解答
     * @param n
     * @return
     */
    public int[] countBits(int n) {
        /*
        计算 0-n 中对应二进制中 1 的个数
         */
        int[] ans = new int[n + 1];
        for (int i = 0; i <= n ; i++) {
            ans[i] = Integer.bitCount(i);
        }
        return ans;
    }

    public int[] countBits_dp(int n) {
        /*
        计算 i 的 1 bit 位数，如果存在 0 <= j < i，且 i 和 和 j 相比，i 的二进制只多了一个 1，还需要再计算 i 的 1 bit 位数吗？不需要
        如果知道 小于等于 i 的最大 2 的幂，那么 bit[i] = bit[i-highBit] + 1;
         */
        int[] ans = new int[n + 1];
        int highBit = 0;
        for (int i = 0; i <= n ; i++) {
            if ((i & (i - 1)) == 0) {
                highBit = i;
            }
            ans[i] = ans[i - highBit] + 1;
        }
        return ans;
    }
}

File: src/main/java/com/oycm/algorithm/e/enum_right_log_left/basic/Solution_17.scala
package com.oycm.algorithm.e.enum_right_log_left.basic

object Solution_17 {

  /**
   * 2364. 统计坏数对的数目 1622
   * https://leetcode.cn/problems/count-number-of-bad-pairs
   *
   * 如果 i < j 且 j - i != nums[j] - nums[i], 称 (i, j) 为 坏数对
   *
   * @param nums 整数数组
   * @return 求 nums 中 坏数对 的数
   */
  def countBadPairs(nums: Array[Int]): Long = {
    /*
    怎么找 j - i = nums(j) - nums(i) 的数量？
    nums(j) - j = nums(i) - i 的数量
    n * (n - 1) / 2 - 好数对 的数量就是答案

     */
    val n = nums.length
    val cnt = n.toLong * (n - 1) / 2
    // 好数对
    var ans = 0L
    val map = scala.collection.mutable.Map[Int, Int]()
    for (j <- nums.indices) {
      // 枚举 j，查找 i
      ans += map.getOrElse(nums(j) - j, 0)
      map(nums(j) - j) = map.getOrElse(nums(j) - j, 0) + 1
    }

    cnt - ans
  }

  def main(args: Array[String]): Unit = {
    println(countBadPairs(Array(4, 1, 3, 3)))
    println(countBadPairs(Array(1, 2, 3, 4, 5)))
    println(countBadPairs(Array(1)))
  }

}
